### PR TITLE
Integrate scikit-learn's `set_output` method into `TransactionEncoder`

### LINKED ISSUE #1085: Integrate scikit-learn's `set_output` method into `TransactionEncoder`
# Describe the workflow you want to enable
In `scikit-learn` [version 1.2](https://scikit-learn.org/1.2/whats_new/v1.2.html#id8), [the `set_output` API was introduced](https://scikit-learn.org/stable/auto_examples/miscellaneous/plot_set_output.html). I would like to expose the API inside of the [`mlxtend.preprocessing.transactionencoder.TransactionEncoder`](https://rasbt.github.io/mlxtend/api_subpackages/mlxtend.preprocessing/#transactionencoder) class. This would allow the user to set the output of :method:`TransactionEncoder.fit_transform` and :method:`TransactionEncoder.transform` to a [`pandas.DataFrame`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html) by default, rather than having to manually create the object after transformation.

# Describe your proposed solution
My proposed solution is to define the [:method:`get_feature_names_out`](https://scikit-learn.org/stable/glossary.html#term-get_feature_names_out) in :class:`TransactionEncoder` as this is required to expose the :method:`set_output`. See [:class:`TransformerMixin`](https://scikit-learn.org/stable/modules/generated/sklearn.base.TransformerMixin.html#sklearn.base.TransformerMixin) and [Developer API for `set_output`](https://scikit-learn.org/stable/developers/develop.html#developer-api-set-output) for more details.

# Describe alternatives you've considered, if relevant
Continue using the method described in the [User Guide](https://rasbt.github.io/mlxtend/user_guide/preprocessing/TransactionEncoder/)—convert the output of the transformer to a `pandas.DataFrame` manually.

# Additional context
- This would require the minimum version of `scikit-learn` to increase from 1.0.2 to 1.2.2.
- I'm willing to take on the PR for this work.

### PR DESCRIPTION
**Pull request by it176131**
### Description
This defines the [:method:`get_feature_names_out`](https://scikit-learn.org/stable/glossary.html#term-get_feature_names_out) in :class:`TransactionEncoder` to expose the :method:set_output.

### Related issues or pull requests
Fixes #1085

### Pull Request Checklist

- [x] Added a note about the modification or contribution to the `./docs/sources/CHANGELOG.md` file (if applicable)
  - See 1c4d3284f29112457cbcd7548f3bb2483a6fa81f
- [x] Added appropriate unit test functions in the `./mlxtend/*/tests` directories (if applicable)
  - See 45cb6cd5d86fcd037107e318225d722bd98d5668, b21bb21e8461225f455e8ea19c5f9cb948e16f29, and 1a02dd5aa8c935f65b3951024c6e75cf86fa081b
- [x] Modify documentation in the corresponding Jupyter Notebook under `mlxtend/docs/sources/` (if applicable)
  - See 35dbeef60de441609d3c0d1d21d6c928e7c16daf
- [x] Ran `PYTHONPATH='.' pytest ./mlxtend -sv` and make sure that all unit tests pass (for small modifications, it might be sufficient to only run the specific test file, e.g., `PYTHONPATH='.' pytest ./mlxtend/classifier/tests/test_stacking_cv_classifier.py -sv`)
  - 20 preexisting unit tests did not pass before work was started. I added two new unit tests to cover the work requested. Both new tests pass and the new work does not affect the prior tests.
- [x] Checked for style issues by running `flake8 ./mlxtend`
  - Prior work outside of this PR does not pass flake8 v7.0.0 guidelines. See mlxtend/feature_selection/column_selector.py:81

### CODE REVIEW

**Comment by rasbt** [comment]
Thanks a lot for the PR, really appreciate it and hope to review it in the upcoming days!

**Comment by it176131** [comment]
> Thanks a lot for the PR, really appreciate it and hope to review it in the upcoming days!

Sure thing! I wasn't sure what to put regarding newer version release dates so I bumped the patch number by one and set the release date to "TBD". Lmk if I should change anything.

**Comment by it176131** [comment]
Hey @rasbt it looks like the pipeline checks keep failing for files unrelated to this PR. Should I log a separate issue and open a new PR to fix those too?

**Comment by rasbt** [comment]
I'd say as long as the tests for the new feature pass then it should be ok. There have been some other tests that have been failing in some submodules for several months due to certain software version updates and minor precision differences I think. I haven't had time to investigate yet.

From a quick look, there seems to be a more major problem though:

```
      AttributeError: module 'pkgutil' has no attribute 'ImpImporter'. Did you mean: 'zipimporter'?
      [end of output]
```

Maybe that's related to Python 3.12. That's definitely something worth fixing so we can find out whether the relevant tests for this PR pass. Maybe doing it in a separate branch first may make sense so the PR doesn't become too cluttered. I also changed the setting here in hope tests will now run automatically each time the PR is updated.

Screenshot 2024-03-28 at 8 35 52 AM

And thanks again for your efforts ... I wish I could be more responsive but it's a busy week

**Comment by it176131** [comment]
> I'd say as long as the tests for the new feature pass then it should be ok. There have been some other tests that have been failing in some submodules for several months due to certain software version updates and minor precision differences I think. I haven't had time to investigate yet.
> 
> From a quick look, there seems to be a more major problem though:
> 
> ```
>       AttributeError: module 'pkgutil' has no attribute 'ImpImporter'. Did you mean: 'zipimporter'?
>       [end of output]
> ```
> 
> Maybe that's related to Python 3.12. That's definitely something worth fixing so we can find out whether the relevant tests for this PR pass. Maybe doing it in a separate branch first may make sense so the PR doesn't become too cluttered. I also changed the setting here in hope tests will now run automatically each time the PR is updated.
> 
> Screenshot 2024-03-28 at 8 35 52 AM
> And thanks again for your efforts ... I wish I could be more responsive but it's a busy week

I'll do some digging. If I can replicate the issue and come up with a solution I'll submit it in a separate PR.

**Comment by rasbt** [comment]
No worries, it should be fixed now via #1089 (except for the transaction encoder test, but that's something we can address in this PR)

**Comment by it176131** [comment]
> No worries, it should be fixed now via #1089 (except for the transaction encoder test, but that's something we can address in this PR)

I noticed that scikit-learn 1.1.3 is being [installed in the github workflow](https://github.com/rasbt/mlxtend/pull/1089/files#diff-82e2d176e09c6e8b5b0942389f985f2b18fb29e41a2e36dfb1c3dd7eef912320R34). Can we bump it to 1.2.2 as that is required for `set_output` to work?

**Comment by rasbt** [comment]
Yes, please feel free to bump it up. We probably need to fix some other places that have not been adjusted for the most recent version though

**Comment by it176131** [comment]
Running `test_inverse_transform` locally, it appears the error is raised when `np.array(data_sorted)` is called. This is because the `data_sorted` is list of lists where the nested lists may have varying lengths (same for `oht.inverse_transform(expect)`). I see two potential ways around this:
1. Use a simpler assertion like `assert data_sorted == oht.inverse_transform(expect)`
2. Add `dtype="object"` to the `np.array` constructors

Both result in the test passing. I'm in favor of the first option as it doesn't require any changes to the `data_sorted` _and_ the expected output type is a `list` rather than an `np.ndarray`.

**Comment by it176131** [comment]
Turns out `scikit-learn` version 1.2.2 had a bug in the `set_output` API that failed when the input was not a `pandas.DataFrame`. See [this PR](https://github.com/scikit-learn/scikit-learn/pull/27044) for details. The fix came in `scikit-learn` version 1.3.1. Bumping the `scikit-learn` version to it fixes the issue. Commit(s) to follow.

**Comment by it176131** [comment]
Of course bumping the version results in more failed tests 🤦

I'm going to make a separate PR to handle the `scikit-learn` version bump and the failed unit tests. THEN maybe this will work. Didn't realize it was going to take so much work 😅

**Comment by rasbt** [comment]
Arg sorry. Yeah, a sklearn bump was overdue but I recently didn't have the time to look into it since I wasn't using the affected features. If this is too much work, don't worry about it, I can understand if you want to drop this. I could revisit the version bump in the upcoming weeks then, address this, and then merge your PR once it's addressed.

**Comment by it176131** [comment]
@rasbt looks like all checks passed. Ready to merge 😎

File: mlxtend/preprocessing/transactionencoder.py [hunk 1]
```diff
@@ -181,3 +182,16 @@ def inverse_transform(self, array):
     def fit_transform(self, X, sparse=False):
         """Fit a TransactionEncoder encoder and transform a dataset."""
         return self.fit(X).transform(X, sparse=sparse)
+
+    def get_feature_names_out(self):
+        """Used to get the column names of pandas output.
+
+        This method combined with the TransformerMixin exposes the
```
Comment (rasbt) [inline mlxtend/preprocessing/transactionencoder.py:189]:
```suggestion
        This method combined with the `TransformerMixin` exposes the
```

File: mlxtend/preprocessing/transactionencoder.py [hunk 2]
```diff
@@ -181,3 +182,16 @@ def inverse_transform(self, array):
...
+        set_output API to the TransactionEncoder. This allows the user
```
Comment (rasbt) [inline mlxtend/preprocessing/transactionencoder.py:190]:
```suggestion
        set_output API to the `TransactionEncoder`. This allows the user
```

File: mlxtend/preprocessing/transactionencoder.py [hunk 3]
```diff
@@ -181,3 +182,16 @@ def inverse_transform(self, array):
...
+        to set the transformed output to a pandas.DataFrame by default.
```
Comment (rasbt) [inline mlxtend/preprocessing/transactionencoder.py:191]:
```suggestion
        to set the transformed output to a `pandas.DataFrame` by default.
```

**Review by rasbt — APPROVED** [review]
This looks great. Thanks so much for the efforts! And sorry again about the unit test hassles!

**Pull request merged**